File: frontend/src/api/notifications.ts
import apiClient from './client'
import { Notification } from '../types/notification'

export const getNotifications = async (): Promise<Notification[]> => {
  const response = await apiClient.get<Notification[]>('/api/v1/notifications/my')
  return response.data
}

export const markAsRead = async (id: string): Promise<Notification> => {
  const response = await apiClient.patch<Notification>(`/api/v1/notifications/${id}/read`)
  return response.data
}

export const markAllAsRead = async (): Promise<void> => {
  await apiClient.patch('/api/v1/notifications/read-all')
}

File: backend/app/routers/notifications.py
from uuid import UUID
from typing import List

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from app.database import get_db
from app.schemas.notification import NotificationResponse
from app.services.notification_service import (
    get_user_notifications,
    mark_notification_as_read,
    mark_all_notifications_as_read,
)
from app.dependencies import get_current_user
from app.models.user import User

router = APIRouter()


@router.get("/my", response_model=List[NotificationResponse])
def get_my_notifications(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    notifications = get_user_notifications(db, current_user.id)
    return [NotificationResponse.model_validate(n) for n in notifications]


@router.patch("/{notification_id}/read", response_model=NotificationResponse)
def mark_as_read(
    notification_id: UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    notification = mark_notification_as_read(db, notification_id, current_user.id)
    if not notification:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Notification not found")
    return NotificationResponse.model_validate(notification)


@router.patch("/read-all")
def mark_all_as_read(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    mark_all_notifications_as_read(db, current_user.id)
    return {"message": "All notifications marked as read"}

